src/attached_probe.cpp

```diff
@@ -982,267 +1113,172 @@ Result<> AttachedProbe::attach_usdt(BPFfeature &feature)
+  int perf_event_fd = bpf_attach_uprobe(progfd,
+                                        attachtype(probe.type),
+                                        eventname(probe, offset).c_str(),
+                                        probe.path.c_str(),
+                                        offset,
+                                        pid.has_value() ? *pid : -1,
                                         semaphore_offset);
 
   if (perf_event_fd < 0) {
-    if (pid_.has_value())
+    if (pid.has_value())
       return make_error<AttachError>("Does PID exist? PID: " +
-                                     std::to_string(*pid_));
+                                     std::to_string(*pid));
 
     return make_error<AttachError>();
   }
 
-  perf_event_fds_.push_back(perf_event_fd);
-  return OK();
-}
-
-Result<> AttachedProbe::attach_tracepoint()
-{
-  int perf_event_fd = bpf_attach_tracepoint(prog_.fd(),
-                                            probe_.path.c_str(),
-                                            eventname().c_str());
-
-  if (perf_event_fd < 0 && probe_.name == probe_.orig_name) {
-    return make_error<AttachError>();
-  }
+  return [perf_event_fd,
+          usdt_sem_cleanup = std::move(*sem_up_res),
+          probe_name = probe.name]() {
+    if (bpf_close_perf_event_fd(perf_event_fd))
+      LOG(WARNING) << "failed to close perf event FD for probe: " << probe_name;
 
-  perf_event_fds_.push_back(perf_event_fd);
-  return OK();
+    usdt_sem_cleanup();
+  };
 }
 
-Result<> AttachedProbe::attach_profile()
+Result<std::unique_ptr<AttachedProbe>> AttachedProbe::make(
+    Probe &probe,
+    const BpfProgram &prog,
+    std::optional<int> pid,
+    BPFtrace &bpftrace,
+    bool safe_mode)
 {
-  int group_fd = -1;
-
-  uint64_t period, freq;
-  if (probe_.path == "hz") {
-    period = 0;
-    freq = probe_.freq;
-  } else if (probe_.path == "s") {
-    period = probe_.freq * 1e9;
-    freq = 0;
-  } else if (probe_.path == "ms") {
-    period = probe_.freq * 1e6;
-    freq = 0;
-  } else if (probe_.path == "us") {
-    period = probe_.freq * 1e3;
-    freq = 0;
-  } else {
-    return make_error<AttachError>("invalid profile path \"" + probe_.path +
-                                   "\"");
-  }
-
-  std::vector<int> cpus = util::get_online_cpus();
-  for (int cpu : cpus) {
-    int perf_event_fd = bpf_attach_perf_event(prog_.fd(),
-                                              PERF_TYPE_SOFTWARE,
-                                              PERF_COUNT_SW_CPU_CLOCK,
-                                              period,
-                                              freq,
-                                              pid_.has_value() ? *pid_ : -1,
-                                              cpu,
-                                              group_fd);
-
-    if (perf_event_fd < 0) {
-      return make_error<AttachError>();
+  LOG(V1) << "Trying to attach probe: " << probe.orig_name;
+  int progfd = prog.fd();
```

Comment:
Can you pls propagate down `prog` into the attach_* functions? We'll need it to do libbpf attachment